### PR TITLE
fix(challenge):db conn is not closed

### LINKED ISSUE #449: postgres connection might have memory leak issue
`OpenConnections always equals ConnectionsInUse`, it might be a memory leak. a db connection should be closed and backed to connection pool after command is executed.

Screen Shot 2021-12-02 at 08 05 55

### PR DESCRIPTION
**Pull request by cnlangzi**
### Fixes
- fixed https://github.com/0chain/blobber/issues/449

### Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/blobber/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

### Associated PRs (Link as appropriate):
-

### CODE REVIEW

File: code/go/0chain.net/blobbercore/challenge/challenge.go [hunk 1]
```diff
@@ -41,58 +40,67 @@ func syncOpenChallenges(ctx context.Context) {
...
+
+	// it is not First and Next challenge
+	if !isValid {
+		logging.Logger.Error("[challenge]Challenge chain is not valid")
+		return
+	}
+
+	logging.Logger.Info("[challenge]add: ", zap.String("challenge_id", nextChallenge.ChallengeID))
+	nextChallenge.Status = Accepted
+	nextChallenge.CreatedAt = common.ToTime(nextChallenge.Created)
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/challenge/challenge.go:95]: Shouldn't the UpdatedAt = common.ToTime(...), and the CreatedAt keep unchanged?

File: code/go/0chain.net/blobbercore/challenge/entity.go [hunk 2]
```diff
@@ -107,6 +111,11 @@ func unMarshalField(stringObj datatypes.JSON, dest interface{}) error {
 }
 
 func (cr *ChallengeEntity) Save(ctx context.Context) error {
+	db := datastore.GetStore().GetTransaction(ctx)
+	return cr.SaveWith(db)
+}
+
+func (cr *ChallengeEntity) SaveWith(db *gorm.DB) error {
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/challenge/entity.go:118]: Seems the db is the only option for SaveWith, if so we should still use it the previous way to keep it simple

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/blobbercore/challenge/challenge.go (hunk 1 above)
Comment (cnlangzi) [reply code/go/0chain.net/blobbercore/challenge/challenge.go:95]: No. it is new challenge from blockchain. so CreatedAt means when it is generated on blockchain, not is added on db.

File: code/go/0chain.net/blobbercore/challenge/entity.go (hunk 2 above)
Comment (cnlangzi) [reply code/go/0chain.net/blobbercore/challenge/entity.go:118]: NO. they are different.  it is a db session with heavy transaction if it is got from `GetTransaction(ctx)`. It is a normal db session without transaction in `SaveWith`.  that is why It is added. but `Save` is still used on other code with built-in transaction. it is not easy to refactor it. 
On blobber, transaction always be started in `WithConnection` and `WithReadonlyConnection` for a new incoming request. it is not good .but it is really difficult to refactor all of them.

**Comment by cnlangzi** [comment]
@peterlimg all comments are updated. please check it again. thanks.

**Review by peterlimg — APPROVED** [review]
lgtm

**Pull request merged**